prov/sm2/src/sm2_progress.c

```diff
@@ -289,12 +435,98 @@ static int sm2_alloc_xfer_entry_ctx(struct sm2_ep *ep,
 static int sm2_progress_recv_msg(struct sm2_ep *ep,
 				 struct sm2_xfer_entry *xfer_entry)
 {
+	struct sm2_av *av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct fid_peer_srx *peer_srx = sm2_get_peer_srx(ep);
 	struct fi_peer_rx_entry *rx_entry;
+	struct sm2_cma_data *cma_data;
 	struct sm2_av *sm2_av;
 	fi_addr_t addr;
+	uint64_t comp_flags;
 	int ret;
 
+	/* TODO - Switch on protocol before switching on op to avoid messy
+	 * checks like this */
+
+	/* We don't want to look for rx entries for xfer entries with
+	 * FI_SM2_CMA_HOST_TO_DEV and FI_SM2_CMA_HOST_TO_DEV_ACK flags*/
+	if (xfer_entry->hdr.proto == sm2_proto_cma) {
+		if (xfer_entry->hdr.proto_flags & FI_SM2_CMA_HOST_TO_DEV) {
+			/* Sender received IPC handle from receiver and needs to
+			 * do a device memcpy, so skip rx entry lookup */
+			ret = sm2_cma_hmem_memcpy(ep, xfer_entry);
+			if (ret) {
+				FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+					"Device memcpy in host to device IPC "
+					"protocol failed\n");
+				if (xfer_entry->hdr.proto_flags &
+				    FI_SM2_UNEXP) {
+					xfer_entry->hdr.op_flags &=
+						~FI_DELIVERY_COMPLETE;
+					sm2_fifo_write_back(ep, xfer_entry);
+				} else {
+					smr_freestack_push(
+						sm2_freestack(ep->self_region),
+						xfer_entry);
+				}
+				goto out;
+			}
+
+			ret = sm2_complete_tx(
+				ep, (void *) xfer_entry->hdr.context,
+				xfer_entry->hdr.op, xfer_entry->hdr.op_flags);
+
+			if (ret) {
+				FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+					"Unable to process tx completion\n");
+			}
+
+			sm2_fifo_write_back_ipc_host_to_dev(ep, xfer_entry);
+
+			goto out;
+		} else if (xfer_entry->hdr.proto_flags &
+			   FI_SM2_CMA_HOST_TO_DEV_ACK) {
+			/* Generate receive completion on the receiver because
+			 * the receiver received a IPC device to host ack
+			 * message */
+			cma_data =
+				(struct sm2_cma_data *) xfer_entry->user_data;
+
+			comp_flags = sm2_rx_cq_flags(xfer_entry->hdr.op,
+						     cma_data->rx_entry->flags,
+						     xfer_entry->hdr.op_flags);
+
+			ret = sm2_complete_rx(
+				ep, cma_data->rx_entry->context,
+				xfer_entry->hdr.op, comp_flags,
+				xfer_entry->hdr.size, cma_data->iov[0].iov_base,
+				xfer_entry->hdr.sender_gid, xfer_entry->hdr.tag,
+				xfer_entry->hdr.cq_data);
+
+			if (ret) {
+				FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+					"Unable to process rx completion\n");
+			}
+
+			if (xfer_entry->hdr.proto_flags & FI_SM2_UNEXP) {
+				/* The xfer_entry was actually allocated on the
+				 * receiver side, so we just push it back */
+				smr_freestack_push(
+					sm2_freestack(ep->self_region),
+					xfer_entry);
```

Comment:
See previous comment. I don't think we should ever have an unexpected message in this path.